tests/test_filer_ftp_pytest.py

```diff
@@ -0,0 +1,221 @@
+""" Tests for 'filer.py' FTP functionalities using 'pytest'."""
+
+from unittest import mock
+import ftplib
+import os
+
+from tesk_core.filer import (
+    FTPTransput,
+    Type,
+    ftp_login,
+    ftp_upload_file,
+    ftp_download_file,
+    ftp_check_directory,
+    ftp_make_dirs
+)
+
+
+def test_ftp_login(mocker):
+    """ Ensure ftp_login detects ftp credentials and properly calls
+        ftplib.FTP.login."""
+
+    conn = mocker.patch('ftplib.FTP')
+    mock_login = mocker.patch('ftplib.FTP.login')
+    with mock.patch.dict(
+            'os.environ',
+            {
+                'TESK_FTP_USERNAME': 'test',
+                'TESK_FTP_PASSWORD': 'test_pass',
+            }
+    ):
+        ftp_login(conn, None, None)
+        mock_login.assert_called_with('test', 'test_pass')
+
+
+def test_ftp_upload_file_error(mocker, caplog):
+    """ Ensure that upon upload error, ftp_upload_file behaves correctly."""
+
+    conn = mocker.patch('ftplib.FTP')
+    mocker.patch('ftplib.FTP.storbinary', side_effect=ftplib.error_reply)
+    assert 1 == ftp_upload_file(conn,
+                                'tests/test_filer.py',
+                                '/home/tesk/test_copy.py')
+    assert 'Unable to upload file' in caplog.text
+
+
+def test_ftp_download_file_error(mocker, caplog):
+    """ Ensure that upon download error, ftp_download_file behaves correctly.
+    """
+
+    conn = mocker.patch('ftplib.FTP')
+    mocker.patch('ftplib.FTP.retrbinary', side_effect=ftplib.error_perm)
+    with mock.patch('builtins.open', mock.mock_open(), create=False) as m:
+        assert 1 == ftp_download_file(conn,
+                                      'test_filer_ftp_pytest.py',
+                                      'test_copy.py')
+        assert 'Unable to download file' in caplog.text
+
+
+def test_ftp_download_file_success(mocker, caplog):
+    """ Ensure that upon successful download, the local destination file has
+        been created."""
+
+    conn = mocker.patch('ftplib.FTP')
+    mock_retrbin = mocker.patch('ftplib.FTP.retrbinary')
+    with mock.patch('builtins.open', mock.mock_open(), create=False) as m:
+        assert 0 == ftp_download_file(conn,
+                                      'test_filer_ftp_pytest.py',
+                                      'test_copy.py')
+
+        mock_retrbin.assert_called_with(
+            "RETR " + "test_filer_ftp_pytest.py",
+            mock.ANY
+        )
+        # Since we want to avoid file creation in testing and we're using
+        # 'create=False', we cannot check whether a file exists or not (but
+        # it's not really necessary since we can assert that the necessary
+        # functions have been invoked.
+        # assert os.path.exists('test_copy.py')
+

```

Comment:
Could you please explain that a bit further?

Reply:
I put the comment on too few lines. Similarly to what I have requested for downloading a directory, it would be nice to prove by asserting that the `mock_open` was indeed called for a file that we expected in `test_ftp_download_file_success`.